### PR TITLE
[OptApp] Refactor the Optimization Problem

### PR DESCRIPTION
**Pull request by Igarizza**
**📝 Description**
Refactoring Optimization problem to avoid circular import issue. PR adds ResponseRoutine and MasterControl to the OptimizationProblem to simplifies access to their components.

Please mark the PR with appropriate tags: 
- Api Breaker, Mpi, etc...

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- Added feature X to Y
- Added Foo Application
- Fixed X (#XXXX Reference to issue if apply) 
- etc...

### CODE REVIEW

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py [hunk 1]
```diff
@@ -42,6 +42,8 @@ def __init__(self, model:Kratos.Model, parameters: Kratos.Parameters, optimizati
         parameters.ValidateAndAssignDefaults(self.GetDefaultParameters())
 
         self.master_control = MasterControl() # Need to fill it with controls
+        self._optimization_problem.AddComponent(self.master_control)
+        self._optimization_problem.AddComponent(self.master_control)
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_steepest_descent.py]:
```suggestion
        self._optimization_problem.AddComponent(self.master_control)
```

**Review by sunethwarna — COMMENTED** [review]
Thanks @Igarizza . I would implement the followings:
- `OptimizationProblem::Get/Has/RemoveResponseRoutine` methods and `GetListOfResonseRoutine` method
- `OptimizationProblem::Get/Has/RemoveMasterControl` methods and `GetListOfMasterControls` method
- Need to change function signatures in `OptimizationProblem` to add `ResponseRoutine` and `MasterControl` for python hinting support

to be consistent.

**Comment by Igarizza** [comment]
> Thanks @Igarizza . I would implement the followings:
> 
>     * `OptimizationProblem::Get/Has/RemoveResponseRoutine` methods and `GetListOfResonseRoutine` method
> 
>     * `OptimizationProblem::Get/Has/RemoveMasterControl` methods and `GetListOfMasterControls` method
> 
>     * Need to change function signatures in `OptimizationProblem` to add `ResponseRoutine` and `MasterControl` for python hinting support
> 
> 
> to be consistent.

Done!

File: applications/OptimizationApplication/python_scripts/utilities/optimization_problem.py [hunk 2]
```diff
@@ -28,7 +30,9 @@ def __init__(self, echo_level: int = 0) -> None:
         self.__components: 'dict[Any, dict[str, Any]]' = {
             ResponseFunction: {},
             ExecutionPolicy: {},
-            Control: {}
+            Control: {},
+            MasterControl:{},
+            ResponseRoutine:{}
```
Comment (talhah-ansari) [inline applications/OptimizationApplication/python_scripts/utilities/optimization_problem.py:35]: The hints in line 48, 55, 68 could be extended to include MasterControl and ResponseRoutine.

Comment (Igarizza) [reply applications/OptimizationApplication/python_scripts/utilities/optimization_problem.py:35]: Thnx!

**Review by sunethwarna — APPROVED** [review]

**Review by sunethwarna — DISMISSED** [review]
Thanks @igarizza

**Review by talhah-ansari — DISMISSED** [review]
Thanks @Igarizza

**Review by sunethwarna — DISMISSED** [review]

**Review by talhah-ansari — APPROVED** [review]
Great, thanks @Igarizza

**Pull request merged**